### PR TITLE
Basic UI flow for creating an invoice account 

### PR DESCRIPTION
**Pull request by Stephan-Hough**
-- new module added for invoice accounts
-- basic ui flow for creating an invoice account
-- form.njk updated to accept caption in the title
-- session-forms changed to accept more query params for invalid forms

### CODE REVIEW

File: src/internal/modules/invoice-accounts/forms/add-fao.js [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+ *
+ * @param {Object} request The Hapi request object
+ * @param {Boolean}  selected value used to determine what radio option should be checked
+  */
+const addFaoForm = (request, selected) => {
+  const { csrfToken } = request.view;
+  const action = '/invoice-accounts/create/add-fao';
+  // when the form is validated the regionId and companyId is set to ''
+  const regionId = (request.params.regionId) ? request.params.regionId : '';
+  const companyId = request.params.companyId ? request.params.companyId : '';
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/forms/add-fao.js:18]: done thanks

File: src/internal/modules/invoice-accounts/lib/data-service.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+const getCompanyAddresses = async (companyId) => {
+  const companyAddresses = await services.water.companies.getAddresses(companyId);
+  // get the unique list of addresses
+  const uniqueIds = new Set(companyAddresses.map(row => row.address.id));
+  const uniqueAddresses = Array.from(uniqueIds).map(id => {
+    const { address } = companyAddresses.find(row => row.address.id === id);
+    return address;
+  });
+
+  return uniqueAddresses;
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/lib/data-service.js:30]: Done thanks

File: src/internal/modules/invoice-accounts/routes.js [hunk 3]
```diff
@@ -0,0 +1,146 @@
...
+          }
+        },
+        validate: {
+          params: {
+            regionId: VALID_GUID,
+            companyId: VALID_GUID
+          },
+          query: {
+            form: Joi.string().optional(),
+            redirectPath: Joi.string().optional(),
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/routes.js:32]: done thanks

File: src/internal/views/nunjucks/invoice-accounts/check-details.njk [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+                {% if address.addressLine2 %}
+                  {{ address.addressLine2 }}<br>
+                {% endif %}
+                {% if address.addressLine3 %}
+                  {{ address.addressLine3 }}<br>
+                {% endif %}
+                {% if address.addressLine4 %}
+                  {{ address.addressLine4 }}<br>
+                {% endif %}
+                {{ address.postcode }}
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/views/nunjucks/invoice-accounts/check-details.njk:45]: done thanks

File: src/internal/views/nunjucks/invoice-accounts/check-details.njk [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+      </div>
+    </div>
+    <form method="POST" action="/invoice-accounts/create/save-details" novalidate="">
+      <input type="hidden" name="regionId" value={{regionId}}>
+      <input type="hidden" name="companyId" value={{companyId}}>
+      <input type="hidden" name="csrf_token" value={{csrfToken}}>
+      <button class="govuk-button" data-module="govuk-button">
+        Continue
+      </button>      
+    </form>
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/views/nunjucks/invoice-accounts/check-details.njk:82]: done

File: src/internal/modules/invoice-accounts/controller.js [hunk 6]
```diff
@@ -0,0 +1,176 @@
...
+    dataService.sessionManager(request, regionId, companyId, { agent: agentId });
+    return 'select-address';
+  }
+};
+
+const postCompany = async (request, h) => {
+  const { regionId, companyId } = request.payload;
+  const company = await dataService.getCompany(companyId);
+  const schema = selectCompanyFormSchema(request.payload);
+  const form = forms.handleRequest(selectCompanyForm(request, company, company), request, schema);
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/controller.js]: done thanks

File: src/internal/modules/invoice-accounts/controller.js [hunk 7]
```diff
@@ -0,0 +1,176 @@
...
+  return h.view('nunjucks/form', {
+    ...request.view,
+    caption: getCaption(session.viewData.licenceNumber),
+    pageTitle: 'Do you need to add an FAO?',
+    back: '/manage',
+    form: sessionForms.get(request, addFaoForm(request, !!session.contact))
+  });
+};
+
+const processFaoFormData = (request, regionId, companyId, addFao) => {
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/controller.js]: done

File: src/internal/modules/invoice-accounts/controller.js [hunk 8]
```diff
@@ -0,0 +1,176 @@
...
+  const { regionId, companyId } = request.params;
+  const session = dataService.sessionManager(request, regionId, companyId);
+  if (Object.keys(session).length === 0 && session.constructor === Object) {
+    throw boom.notFound('Session data not found');
+  }
+  const [ company, addresses ] = await Promise.all([
+    dataService.getCompany(companyId),
+    dataService.getCompanyAddresses(companyId)
+  ]);
+  const [ address ] = addresses.filter(address => (address.addressId = session.address.addressId));
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/controller.js]: done

File: src/internal/modules/invoice-accounts/controller.js [hunk 9]
```diff
@@ -0,0 +1,176 @@
...
+  return h.postRedirectGet(form, urlJoin('/invoice-accounts/create/', regionId, companyId), { redirectPath });
+};
+
+const getAddress = async (request, h) => {
+  const { regionId, companyId } = request.params;
+  const addresses = await dataService.getCompanyAddresses(companyId);
+  // get the session data to check if the address has been set and if it is new or existing
+  const session = dataService.sessionManager(request, regionId, companyId);
+  // @TODO this might need a mapper to map the session address data to the Company address shape passed to the form
+  if (session.address && session.address.addressId === tempId) { addresses.push(session.address); }
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/controller.js:59]: The UI needs an id for it to work so a temporary id is assigned and then removed again before posting to the water service.

File: src/internal/modules/invoice-accounts/controller.js [hunk 10]
```diff
@@ -0,0 +1,176 @@
...
+    regionId,
+    company,
+    address });
+};
+
+const postCheckDetails = async (request, h) => {
+  const { regionId, companyId } = request.payload;
+  const session = dataService.sessionManager(request, regionId, companyId);
+  // TODO default start date added here - might need to create a screen for the user to select a date
+  session.startDate = moment().format('YYYY-MM-DD');
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/controller.js:133]: I think we should change this when we connect this flow to the rest of the Charge Info process. At the moment to test this ticket in isolation it requires a date to send it to the water service to post the invoice account object.

File: src/internal/modules/invoice-accounts/forms/select-company.js [hunk 11]
```diff
@@ -0,0 +1,79 @@
+'use strict';
+
+const Joi = require('@hapi/joi');
+const { formFactory, fields } = require('shared/lib/forms/');
+
+const choices = (company, companySelected) => {
+  const options = [{ value: company.id, label: company.name, hint: company.type }];
+  // if an agent company was previously selected add this to the list of options
+  if (company.id !== companySelected.id) {
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/forms/select-company.js]: done

File: src/internal/modules/invoice-accounts/forms/select-address.js [hunk 12]
```diff
@@ -0,0 +1,72 @@
...
+});
+
+/**
+ * Returns the selected address id for the invoice account
+ * along with the predtermined company id and region id
+ * @param {Object} request The Hapi request object
+ * @param {Array} addresses Array of all company address
+ * @param {uuid} selectedAddressId address id stored in session data for pre-selected option
+  */
+const selectAddressForm = (request, addresses, selectedAddressId) => {
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/forms/select-address.js:32]: done

File: src/internal/modules/invoice-accounts/lib/data-service.js [hunk 13]
```diff
@@ -0,0 +1,49 @@
+'use-strict';
+
+const hoek = require('@hapi/hoek');
+const services = require('../../../lib/connectors/services');
+
+const sessionManager = (request, regionId, companyId, data = null) => {
+  // get existing session data
+  let sessionData = request.yar.get(`newInvoiceAccountFlow.${regionId}.${companyId}`);
+  // merge the new with old data
+  sessionData = hoek.merge(sessionData || {}, { ...data });
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/lib/data-service.js:10]: yes I can - Thanks :-)

File: src/internal/modules/invoice-accounts/controller.js (hunk 9 above)
[the remark this answers is not in the document]
Comment (DGajewska) [reply src/internal/modules/invoice-accounts/controller.js:59]: Sounds good. I know that you've tested it with the backend, but it's safer to ask than to assume things! Thank you :)

File: src/internal/modules/invoice-accounts/lib/helpers.js [hunk 14]
```diff
@@ -0,0 +1,40 @@
...
+    return 'check-details';
+  }
+};
+
+const getSelectedAddress = async (companyId, session) => {
+  if (session.address.addressId === tempId) {
+    return session.address;
+  } else {
+    const addresses = await dataService.getCompanyAddresses(companyId);
+    const [ selectedAddress ] = addresses.filter(address => (address.addressId === session.address.addressId));
```
Comment (DGajewska) [inline src/internal/modules/invoice-accounts/lib/helpers.js:33]: If you're only looking for 1 result, you should use the `find` method as it returns the first matching result

File: src/internal/modules/invoice-accounts/forms/select-company.js [hunk 15]
```diff
@@ -0,0 +1,76 @@
...
+ * returns the selected company id along with the region and company id
+ *
+ * @param {Object} request The Hapi request object
+ * @param {Object} company The main company for the licence
+ * @param {Object} companySelected The selected company which could be different from the main company
+  */
+const selectCompanyForm = (request, company, companySelected = null) => {
+  const { csrfToken } = request.view;
+  const regionId = request.params.regionId || '';
+  const companyId = request.params.companyId || '';
```
Comment (DGajewska) [inline src/internal/modules/invoice-accounts/forms/select-company.js:42]: Same as above

File: src/internal/modules/invoice-accounts/forms/check-details.js [hunk 16]
```diff
@@ -0,0 +1,24 @@
...
+ * Returns the selected address id for the invoice account
+ * along with the predtermined company id and region id
+ * @param {Object} request The Hapi request object
+ * @param {Array} addresses Array of all company address
+ * @param {uuid} selectedAddressId address id stored in session data for pre-selected option
+  */
+const checkDetailsForm = (request) => {
+  const { csrfToken } = request.view;
+  const regionId = request.params.regionId || '';
+  const companyId = request.params.companyId || '';
```
Comment (DGajewska) [inline src/internal/modules/invoice-accounts/forms/check-details.js:16]: Same as above

File: src/internal/modules/invoice-accounts/forms/add-fao.js [hunk 17]
```diff
@@ -12,10 +13,10 @@ const { formFactory, fields } = require('shared/lib/forms/');
   */
 const addFaoForm = (request, selected) => {
   const { csrfToken } = request.view;
-  const action = '/invoice-accounts/create/add-fao';
   // when the form is validated the regionId and companyId is set to ''
-  const regionId = (request.params.regionId) ? request.params.regionId : '';
-  const companyId = request.params.companyId ? request.params.companyId : '';
+  const regionId = request.params.regionId || '';
+  const companyId = request.params.companyId || '';
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/forms/add-fao.js:18]: True thank you

File: src/internal/modules/invoice-accounts/forms/select-address.js [hunk 18]
```diff
@@ -31,9 +31,9 @@ const addressList = (addresses) => addresses.map(row => {
   */
 const selectAddressForm = (request, addresses, selectedAddressId) => {
   const { csrfToken } = request.view;
-  const action = '/invoice-accounts/create/select-address';
-  const regionId = (request.params.regionId) ? request.params.regionId : '';
-  const companyId = request.params.companyId ? request.params.companyId : '';
+  const regionId = request.params.regionId || '';
+  const companyId = request.params.companyId || '';
```
[the remark this answers is not in the document]
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/forms/select-address.js:35]: done

File: src/internal/modules/invoice-accounts/forms/select-company.js (hunk 15 above)
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/forms/select-company.js:42]: done

File: src/internal/modules/invoice-accounts/forms/check-details.js (hunk 16 above)
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/forms/check-details.js:16]: done

File: src/internal/modules/invoice-accounts/forms/add-fao.js [hunk 19]
```diff
@@ -29,7 +31,7 @@ const addFaoForm = (request, selected) => {
         message: 'Select yes if you need to add a person or department as an FAO'
       }
     },
-    choices: [yes, no],
+    choices: [{ value: 'yes', label: 'Yes' }, { value: 'no', label: 'No' }],
```
Comment (DGajewska) [inline src/internal/modules/invoice-accounts/forms/add-fao.js:34]: You can use the `options` you defined above:
`choices: [Object.values(options)]`

File: src/internal/modules/invoice-accounts/forms/select-address.js [hunk 20]
```diff
@@ -3,24 +3,28 @@
...
-    address.addressLine3,
-    address.addressLine4,
-    address.town,
+    titleCase(address.addressLine1),
+    titleCase(address.addressLine2),
+    titleCase(address.addressLine3),
+    titleCase(address.addressLine4),
+    titleCase(address.town),
     address.postcode
   ].filter(item => item !== '' && item !== null);
```
Comment (DGajewska) [inline src/internal/modules/invoice-accounts/forms/select-address.js:16]: I think that you can use the lodash method `isEmpty` here:
`.filter(item => !isEmpty(item));

File: src/internal/modules/invoice-accounts/forms/add-fao.js (hunk 19 above)
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/forms/add-fao.js:34]: done

File: src/internal/modules/invoice-accounts/forms/select-address.js (hunk 20 above)
Comment (Stephan-Hough) [reply src/internal/modules/invoice-accounts/forms/select-address.js:16]: done

**Pull request merged**